### ISSUE TITLE
Support for cross-domain WebDAV access (CORS)

### ISSUE THREAD

**Issue by xMartin** [nextcloud/server#3131]
It would be great if web applications hosted on a different domain could access Nextcloud files via WebDAV. This would enable web app developers to offer users to store personal application data in their Nextcloud without the need to provide a dedicated Nextcloud app.

Currently this is not possible as the necessary CORS headers are not set.

@LukasReschke

**Comment by jancborchardt** [comment]
Also related to this as a future step »Dropbox like file picker javascript integration #2028«

**Comment by jancborchardt** [comment]
@LukasReschke what are the next steps here?

**Comment by doleraj** [comment]
I'd like to help as much as I can with this issue - my end goal is to be able to parse the calendar data out with some JS so my personal blog can detail conferences I've spoken at and places I will be speaking soon. 

Hit me up if you need to bounce ideas for the solution off someone, or need example use cases etc. I don't have a dev environment for nextcloud set up, but if the problem is a shortage of hands I can take a stab at implementing it if I get some guidance.

**Comment by jancborchardt** [comment]
@doleraj most often the reason is a shortage of hands, so your help would be much appreciated! :) For any guidance best join our IRC channel #nextcloud-dev or ask here.

Also cc @perry-mitchell for his work on connecting Nextcloud to web apps: https://nextcloud.com/blog/using-webdav-fs-to-access-files-in-nextcloud/ and cc @nextcloud/javascript

**Comment by perry-mitchell** [comment]
We're currently using webdav+Nextcloud to read and write our password archives both in the browser and in NodeJS, using [this project](https://github.com/perry-mitchell/webdav-fs) (disclaimer: I'm the author). You can see [here](https://github.com/buttercup/buttercup-core/blob/master/source/system/WebDAVDatasource.js#L86) that we write to the storage without any specifics (besides [auth headers](https://github.com/perry-mitchell/webdav-client/blob/master/source/clientFactory.js#L38)).

So far I don't believe we've encountered any CORS issues with access to Nextcloud from within the browser, providing authorization is sent. I don't see why there'd be any problems, but perhaps you might have a specific example in terms of code or cURL etc.? Our library does use a transpiled copy of [`node-fetch`](https://github.com/bitinn/node-fetch) however, which may have some affect on the outcome. It may be that items like native browser `fetch` somehow behaves differently.

**Comment by doleraj** [comment]
The "with authentication" thing may be the trick. I'm trying to get at a public calendar without auth since the code'll be running browser-side.

As an example, hitting `(https://cloud.arthurdoler.com/remote.php/dav/public-calendars/230G65ZAI9PV9KAT?export )` as an actual URL works fine, causing the .ics to be downloaded.
But trying the following: 
```
const url = 'https://cloud.arthurdoler.com/remote.php/dav/public-calendars/230G65ZAI9PV9KAT?export';
        fetch(url).then((response) => {
            console.log(response)
        });
```
causes a CORS error: 
`Fetch API cannot load https://cloud.arthurdoler.com/remote.php/dav/public-calendars/230G65ZAI9PV9KAT?export. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:63342' is therefore not allowed access. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

**Comment by jancborchardt** [comment]
Also cc @georgehrke @tcitworld of the Calendar app for @doleraj's specific question. :)

**Comment by doleraj** [comment]
... oops! I apologize, I'd forgetten that Calendar is actually a separate app.

**Comment by georgehrke** [comment]
@doleraj But public-calendars is part of the dav app which belongs to the server ;) 

@LukasReschke might know more security wise.

**Comment by blizzz** [comment]
We've had a similar issue in bookmarks and ended up having an internal controller being sort of a facade for the public one which incorporates the ``@CORS`` annotation. The issue with the later is that it disables sessions. This solution feels more like an ugly hack, however.

**Comment by tcitworld** [comment]
I discovered that https://github.com/owncloud/core/pull/28457 is merged upstream, maybe we can integrate this.

**Comment by jancborchardt** [comment]
@tcitworld mind opening a pull request to downstream that change? :)

**Comment by jancborchardt** [comment]
@tcitworld mind opening a pull request to downstream that change? :) Or @LukasReschke what do you say?

**Comment by mehrdaad** [comment]
Any updates on this?  
Anyone partitioned in this issue: How to solve Same Origin Policy in Nextcloud?  
What I want to do is send request to WebDAV server from another url. no more, no less.

**Comment by mehrdaad** [comment]
@perry-mitchell hey, could you solve it?

**Comment by perry-mitchell** [comment]
@mehrdaad We have Nextcloud somewhat working in our password management software, but I believe it's a bit buggy still. The NodeJS side works it seems (no credentials or cookies), but when fetching on the client side it get's all sorts of weird. I'll be checking into Nextcloud support again (for the browser) in the coming week, so I'll post back here if I find anything.

**Comment by anmol26s** [comment]
Can I add a sub domain like app.domain.tld to trusted domain for Nextcloud and use that app with webdav without having the CORS issue?

**Comment by perry-mitchell** [comment]
Writing back after some time.. I should've tested Nextcloud again sooner. We're trying to integrate with [Buttercup](https://buttercup.pw) (our password manager), but it doesn't work in the browser. We were originally having CORS issues but when trying to connect to a demo instance (on 12.0.0) we get 503 errors: #7365

As of now, I've not been able to get WebDAV working **at all** in the browser. Connecting from NodeJS works fine every time.

**Comment by PeterNerlich** [comment]
Any update on this? Is there anyone looking into this at the moment? Or is this almost a lost cause?

**Comment by navid9675** [comment]
Same question here, I want to use some features of the Nextcloud on a different client with a different view(limited view), I keep getting 'unauthorized 401' error, any solutions?

**Comment by mabuch** [comment]
Adding the code from https://stackoverflow.com/questions/8719276/cors-with-php-headers/9866124#9866124 to remote.php (just after the already present header-function call) solves the issue for me.

**Comment by dagobit** [comment]
It could be useful to also add the POPFIND to the Access-Control-Allow-Methods, like:

`                header("Access-Control-Allow-Methods: GET, POST, PUT, MOVE, OPTIONS, PROPFIND");
`

**Comment by perry-mitchell** [comment]
Any update on this? Over at [Buttercup](https://github.com/buttercup) we see issues pop up now and then about users not being able to access their Nextcloud servers via WebDAV, which we provide using my library [webdav-client](https://github.com/perry-mitchell/webdav-client). It also has a [couple of issues](https://github.com/perry-mitchell/webdav-client/issues/116) regarding Nextcloud access from the browser.

Are there any plans on supporting this? Will OAuth be the only supported access method via websites? Some official feedback would be much appreciated here.

**Comment by jospoortvliet** [comment]
@rullzer can you have a look? This probably has security implications, but we'd like third parties like buttercup to be able to work with nextcloud of course!

**Comment by perry-mitchell** [comment]
Thanks! This’d be super good. WebDAV really makes things easy, and will definitely do so in a cross platform sense when this works in a browser.

Also open to ideas of using the OAuth system I’ve seen in the dashboard but WebDAV would definitely bring less friction.

**Comment by rullzer** [comment]
I'll look into this. But no time line promises.

**Comment by rullzer** [comment]
So I have been thinking about this while traveling. A few things that come to mind. I need to really dive into the specs. But.

I would say that you can only use webdav via CORS if you are authenticated via an app token.

In a later stage we can then enhance our OAuth API to allow user to more easily obtain OAuth tokens for their apps.

**Comment by perry-mitchell** [comment]
Hmm.. how would such a token be generated? Is there an OAuth flow in Nextcloud (I’m not familiar with any such feature) or do the users need to manually do it in the admin interface?

This is still a bit more work than with WebDAV on other devices, hence it adds more friction and can be a turn off for some looking for an integrated solution. We’ve historically suggested Nextcloud to users due to how easy it’s been but this might be a difficult sell.

WebDAV with Nextcloud works on all devices, except in the browser, using only username and password - it’s just CORS that’s causing an issue.

**Comment by peponi** [comment]
when CORS with a token is working, I'm intrested to build a Nextcloud module for [remotestorage.js](https://github.com/remotestorage/remotestorage.js) with webdav-client

maybe we will find a way then to host pictures in Mastodon in our own cloud accounts and not on the Mastadon pods

**Comment by rullzer** [comment]
@perry-mitchell Improved OAuth flow would be up next then :)

**Comment by memen45** [comment]
Current OAuth flow in Nextcloud does not support browser side clients: #12667 

Looking forward to this feature implementation!

**Comment by perry-mitchell** [comment]
So there's no way to make requests to a Nextcloud server from the browser then? 

Sadly this means that we might have to move on from Nextcloud as a storage backend until browser access is properly supported. CORS is a huge pain but a necessity when building anything with an API these days. I look forward to having such support.. Thank you.

**Comment by rullzer** [comment]
@memen45 @perry-mitchell I'll look into https://tools.ietf.org/html/rfc7636 and see if we can support that

**Comment by zwif** [comment]
I'm also facing an issue with this: I try to access a file via remote.php from an external Web-Application (running in Browser).

The CORS preflight is not successfull. The Nextcloud-Server is expecting an 'Authorization' header for the OPTIONS request (see: `curl -X OPTIONS http://MYSERVER/remote.php/webdav/SOME/FILE.png`).
 But Browser tend to not send the Authorization headers in OPTIONS requests, since that is required by the  [CORS spec](https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0) saying 
> for a cross-origin request with preflight [...] Exclude user credentials

I'm not into the details here but wouldn't it be sufficient to change the server in a way to response to OPTIONS call without requiring Authorization header (but leaving it for GET, POST, ...)?

(Found here: https://stackoverflow.com/questions/52072001/the-authorization-header-is-not-sent-in-options-call-by-the-browser)

**Comment by kinolaev** [comment]
Hello!

I solved this problem with following additions to [recommended nginx config](https://docs.nextcloud.com/server/stable/admin_manual/installation/nginx.html):
```
map $http_origin $access_control_allow {
    default 0;
    https://first-origin.com 1;
    https://second-origin.com 1;
}
map $access_control_allow $access_control_allow_origin {
    default '';
    1 $http_origin;
}
map $access_control_allow $access_control_allow_methods {
    default '';
    1 $http_access_control_request_method;
}
map $access_control_allow $access_control_allow_headers {
    default '';
    1 $http_access_control_request_headers;
}
map $access_control_allow $access_control_expose_headers {
    default '';
    1 'etag';
}
map "$access_control_allow:$request_method" $access_control_return_204 {
    default 0;
    '1:OPTIONS' 1;
}
# ...
server {
    # ...
    location ~ ^\/(?:index|remote|public|cron|core\/ajax\/update|status|ocs\/v[12]|updater\/.+|oc[ms]-provider\/.+)\.php(?:$|\/) {
        add_header Access-Control-Allow-Origin $access_control_allow_origin;
        add_header Access-Control-Allow-Methods $access_control_allow_methods;
        add_header Access-Control-Allow-Headers $access_control_allow_headers;
        add_header Access-Control-Expose-Headers $access_control_expose_headers;
        if ($access_control_return_204) {
            return 204;
        }
        # ...
}
```
@rullzer what do you think about this?

**Comment by michielbdejong** [comment]
Is this link relevant? https://docs.nextcloud.com/server/stable/developer_manual/app/requests/api.html
I realise you would prefer to have CORS headers on the main WebDAV interface, but that doc seems to explain how to develop a NextCloud app that exposes an API with CORS headers on it, right?

**Comment by cougarten** [comment]
@michielbdejong your link is gone for me. I guess you mean this: https://github.com/nextcloud/documentation/blob/3aecb6899b79650a0efc17936f69a2f4ca575288/developer_manual/app/api.rst#modifying-the-cors-headers

(no idea if that helps, but I'd also love to be able to set those headers)

**Comment by thosmos** [comment]
Doing a basic OPTIONS request is failing incorrectly when not including authentication.  The HTTP spec is that OPTIONS should NOT have authentication, so NextCloud is at fault.  

For example, this should succeed:
```
curl -i -X OPTIONS https://mynextcloud.com/remote.php/webdav/MailAttachments/myfile.zip
HTTP/1.1 401 Unauthorized
...
```
If I use a password or token it succeeds, but this is not how it should work...

**Comment by kinolaev** [comment]
Hello, I wrote simple Sabre plugin (not published yet) that adds CORS headers https://gitlab.com/kinolaev/nextcloud-dav-cors. Just download it to apps directory and add array of allowed domains to `dav_cors.origins` in `config.php`.

**Comment by perry-mitchell** [comment]
@kinolaev That's great, thank you for your work!

The issue I still see here is that CORS doesn't help when there's no domain on the client end. Both browser extensions and electron applications, that use WebDAV, do not have an origin as such. Nextcloud should not fault in these scenarios, and CORS should not restrict their access.

**Comment by kinolaev** [comment]
@perry-mitchell, as I understand webextensions [does have origins](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#Host_permissions) and electron apps, I think, can skip CORS check.

**Comment by michielbdejong** [comment]
Will try that out, thanks!

**Comment by pbek** [comment]
Thank you, @kinolaev! Was anyone successful using the app? For me already the `OPTIONS` requests didn't make it to the `OCA\DAV\Connector\Sabre::addPlugin` listener. And even if I waved the `OPTIONS` requests through in the `remote.php` the real request (e.g. `PROPFIND`) didn't make it to the listener neither.

**Comment by kinolaev** [comment]
Hello @pbek,
I updated plugin for nc 19, for nc 18 please use [initial commit](69a7e79fd279a9e2ed41ec5b910f955b88ee775b)

**Comment by pbek** [comment]
Do you mean <https://gitlab.com/kinolaev/nextcloud-dav-cors/-/commit/d47a941635020ef356768affb467698853b8b4a2>, @kinolaev? Thank you, I'll try reverting that. But so far nothing made it into the event listener lambda, the code to add the plugin was never executed for me... Even though the event was fired by Sabr/Dav...

**Comment by kinolaev** [comment]
Yes, you can just replace `beforeMethod:*` with `beforeMethod` for nc 18 (sabre/dav below 4).

**Comment by pbek** [comment]
That really did it! Thank you very much! Should I a make a pull requests to increase your `min-version` to 19? :grin:

**Comment by kinolaev** [comment]
No thanks, I will try to add version check with next few days)

**Comment by pbek** [comment]
Great! Do you care also switching away from the deprecated (as in NC 18) `\OC::$server->getEventDispatcher()` to `\OCP\EventDispatcher\IEventDispatcher`? :grin:

**Comment by kinolaev** [comment]
I updated the plugin:
- `getEventDispatcher` replaced with `IEventDispatcher`
- lastest version works with nextcloud 18 and 19
- fixed phpdoc

**Comment by pbek** [comment]
Awesome, I'll test that tomorrow!

**Comment by pbek** [comment]
Works great, @kinolaev. Thank you very much!

**Comment by pbek** [comment]
There now ist an app in the app store for setting the CORS header:
https://apps.nextcloud.com/apps/webapppassword

It also allows the generation of temporary app passwords to be able access the Nextcloud WebDAV service from inside web applications. 

There is an [example implementation](https://gitlab.tugraz.at/dbp/nextcloud/webapppassword/-/blob/master/index.html) that uses this app to generate a temporary password and make a WebDAV request to a Nextcloud instance ([docker container](https://gitlab.tugraz.at/dbp/nextcloud/webapppassword/-/tree/master/docker) included).

**Comment by pbek** [comment]
Here are some screenshots for https://apps.nextcloud.com/apps/webapppassword:

## Settings
![screenshot](https://user-images.githubusercontent.com/1798101/89513641-1e06f600-d7d5-11ea-834d-3d323e48cf11.png)

## WebDAV File Picker
<img width="1058" alt="screenshot-filepicker" src="https://user-images.githubusercontent.com/1798101/89513605-0fb8da00-d7d5-11ea-9085-40d5e12888d9.png">

## Temporary password generation and WebDAV PROPFIND
<img width="1068" alt="screenshot-webdav" src="https://user-images.githubusercontent.com/1798101/89513610-10517080-d7d5-11ea-928b-597d3c994389.png">

**Comment by MichaelTunnell** [comment]
will this WebAppPassword nextcloud app work with multiple instances so multiple users can have this syncing their Super Productivity data?

**Comment by pbek** [comment]
> work with multiple instances

please define "multiple instances"

> syncing their Super Productivity data?

syncing with what?

**Comment by meyca** [comment]
Screen from https://app.super-productivity.com/#/config

![grafik](https://user-images.githubusercontent.com/408161/104360866-fb120800-5511-11eb-8a00-58ab87a5369e.png)

refers to this thread.

@MichaelTunnell The WebAppPassword app will just allow your Nextcloud instance to "talk" with whatever host your super-productivity-app is running on. As far as I understand, this is not Nextcloud-user dependant. However, there are security implications with this setting!

**Comment by janLo** [comment]
I'm trying to implement Caldav as a Issue-Provider for [SuperProductivity](https://github.com/johannesjo/super-productivity) and ran in the missing CORS headers as the Web and Mobile apps enforce CORS rules. @rullzer, @LukasReschke what is the status here? How likely is it that owncloud/core#28457 gets merged in NC?

**Comment by pbek** [comment]
> I'm trying to implement Caldav as a Issue-Provider for [SuperProductivity](https://github.com/johannesjo/super-productivity) and ran in the missing CORS headers as the Web and Mobile apps enforce CORS rules.

https://apps.nextcloud.com/apps/webapppassword is adding those headers for you for origins you setup.

**Comment by pbek** [comment]
> However, there are security implications with this setting!

CORS is something the browser enforces. As long as the server administrator can set which origins to allow there should be no problems (also if such a setting would be included into NC directly).

**Comment by janLo** [comment]
> https://apps.nextcloud.com/apps/webapppassword is adding those headers for you for origins you setup.

This is not enough for caldav as it seems that the headers have to be allowed as well.

> CORS is something the browser enforces. As long as the server administrator can set which origins to allow there should be no problems (also if such a setting would be included into NC directly).

I think if one provides a Web-API, the CORS settings should be set to sane defaults. Especially as not every user of a NC-Calendar is also the administrator of the instance. Configurability would be great - but the default should always be set to "useable".

**Comment by pbek** [comment]
> This is not enough for caldav as it seems that the headers have to be allowed as well.

Strange, they should be pretty similar. https://gitlab.tugraz.at/dbp/nextcloud/webapppassword/-/blob/master/lib/Connector/Sabre/CorsPlugin.php#L51 was enough for webdav.

> I think if one provides a Web-API, the CORS settings should be set to sane defaults. Especially as not every user of a NC-Calendar is also the administrator of the instance. Configurability would be great - but the default should always be set to "useable".

The CORS headers are a loaded topic. 😅

**Comment by janLo** [comment]
> The CORS headers are a loaded topic. ￼

Yes, but as far as I can see it there is discussion for the last four years, one opstream PR, multiple code snippets, workarounds and ideas but no progress towards a solution or rejection :(.

**Comment by janLo** [comment]
This is what I get with the app mentioned above: `Refused to get unsafe header "DAV"`

**Comment by pbek** [comment]
> I think the DAV-header has to be part of the Access-Control-Expose-Headers - header: https://fetch.spec.whatwg.org/#http-access-control-expose-headers

Did you try adding it to webapppassword to test it?

**Comment by janLo** [comment]
> > I think the DAV-header has to be part of the Access-Control-Expose-Headers - header: https://fetch.spec.whatwg.org/#http-access-control-expose-headers
> 
> Did you try adding it to webapppassword to test it?

I just did. It works if I add dav as exposed header to the app.

**Comment by pbek** [comment]
> > 
> > Did you try adding it to webapppassword to test it?
> 
> I just did. It works if I add dav as exposed header to the app.

What code did you add?

**Comment by janLo** [comment]
https://gitlab.tugraz.at/dbp/nextcloud/webapppassword/-/blob/master/lib/Connector/Sabre/CorsPlugin.php#L54
There I changed
```
$response->addHeader('access-control-expose-headers', 'etag');
```
to
```
$response->addHeader('access-control-expose-headers', 'etag, dav');
```

**Comment by pbek** [comment]
CalDAV is now also allowed in https://apps.nextcloud.com/apps/webapppassword.

**Comment by janLo** [comment]
Thank you. I still hope hat we need get support for CORS directly in NC as this workaround requires - like all the others - admin-access to the NC-Instance.

**Comment by pbek** [comment]
I don't think general "full" CORS access is going to happen by default.

**Comment by Silbad** [comment]
Hello possible to add CORS for OCS API too ?
https://docs.nextcloud.com/server/latest/developer_manual/client_apis/OCS/ocs-share-api.html

**Comment by sunjam** [comment]
Will it be possible, either now or in the future, to use [LoginFlow](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html) to easily and securely access third party applications currently requiring the CORS webapppassword?

**Comment by pbek** [comment]
CORS is only a web browser restriction. To get around the password flow issue we currently use the temporary app passwords we create with https://apps.nextcloud.com/apps/webapppassword.

**Comment by janLo** [comment]
Its still preferable to have a solution available that works for everyone who is not the administrator of their NC instance. @rullzer what is the current status?

**Comment by pbek** [comment]
> Its still preferable to have a solution available that works for everyone who is not the administrator of their NC instance.

In my opinion CORS settings should always be something that an admin needs to setup. That's the whole reason for CORS... To restrict who (what origin) can load resources from a browser. 😁

**Comment by pbek** [comment]
If the admin decides that all origins should be allowed he can still set `*` as origin (for example in https://apps.nextcloud.com/apps/webapppassword). An admin could even go as far and force those headers on Apache/NGINX level, the CORS part of https://apps.nextcloud.com/apps/webapppassword is just for "convenience".

**Comment by janLo** [comment]
It should have sensible defaultts. I think, a good default for a Web-API is to allow to use it from the Web. If an admin wats to forbid this then there will always be options. In the current state its pretty much unuseable for everyone who uses a NC instance in a corporate environment or in a University. On the other hand it encourages people to find workarounds without upstream-support bypassing any security review/audit like local proxies that terminate the encryption and inject the necessary headers.

**Comment by pbek** [comment]
I guess most of the people wouldn't like seeing their private Nextcloud server having a public api, that is accessible from all websites... 😅

**Comment by szaimen** [comment]
> I guess most of the people wouldn't like seeing their private Nextcloud server having a public api, that is accessible from all websites... 😅

Yes! Being able to poke holes through the default secure configuration via the the webapppassword app sounds like the better solution to me 👍

**Comment by janLo** [comment]
> I guess most of the people wouldn't like seeing their private Nextcloud server having a public api, that is accessible from all websites...

Nobody suggested this. But NC has a mechanism already to request app specific access. This could be used to scope the access to origins the app has requested in the first place.

**Comment by janLo** [comment]
That said - CORS is not a protection to have your API exposed to the internet. Its mainly to protect you from yourself by hoping that your browser will not access the API. There is no protection whatsoever on the server itself.

**Comment by pbek** [comment]
> Nobody suggested this. 

I though you suggested it by saying "a good default for a Web-API is to allow to use it from the Web."... So what would be your suggestion (in the case of CORS defaults)? Or were you just referring to the fact that some mechanism like in webapppassword isn't integrated in NC by default? And if that's the case what should be the default setting?

**Comment by pbek** [comment]
> That said - CORS is not a protection to have your API exposed to the internet.

That's correct.

> Its mainly to protect you from yourself by hoping that your browser will not access the API. There is no protection whatsoever on the server itself.

But it would for example make a difference if some webpage with a huge amount of traffic would use your website as CDN for e.g. images you are hosting, wouldn't it? 😁  (that could be the result if you would allow all origins with CORS)

**Comment by neben** [comment]
> Hello possible to add CORS for OCS API too ?

We're currently forced to have a separate nginx configuration for nextcloud to enable OCS access from our web-app. This works for our own nextcloud instance, but it keeps, e.g., students from being able to use our app with the university's  nextcloud instance.

```
map $http_origin $valid_origin {
	hostnames;

	https://foo.example.com	$http_origin;
	https://bar.example.com	$http_origin;
	default 					invalid;
}

server {
[...]
  set $upstream http://nextcloud;

	location / {
		proxy_pass $upstream;
	}

	# set cors for all but /status.php
	location ~ ^/(?!status.php) {
		add_header Access-Control-Allow-Methods $http_access_control_request_method always;
		add_header Access-Control-Allow-Origin	$valid_origin always;
		add_header Access-Control-Allow-Headers $http_access_control_request_headers always;
		add_header Strict-Transport-Security 		"max-age=31536000; includeSubDomains; preload" always;
		if ($request_method = OPTIONS) { return 204; }

		proxy_pass $upstream;
	}
}
```

**Comment by janLo** [comment]
> But it would for example make a difference if some webpage with a huge amount of traffic would use your website as CDN for e.g. images you are hosting, wouldn't it?  (that could be the result if you would allow all origins with CORS)

I honestly don't think that this is any objective of the strict-origin policy. Its more about leaking sensitive information. If you want to protect yourself from high-traffic usage that you should not rely on any client-side measures.

Nextcloud already has a builtin mechanism to request app passwords. This could get extended to add a lost of allowed origins.

**Comment by neben** [comment]
> But it would for example make a difference if some webpage with a huge amount of traffic would use your website as CDN for e.g. images you are hosting, wouldn't it? 😁 (that could be the result if you would allow all origins with CORS)

I'm afraid CORS doesn't help with that. You can always embed cross-origin images, you're just limited in the way you can access/manipulate that image data with JS. You can try it, share an image file in nextcloud, and embed that image on a different domain using the share URL + '/preview':
```
<img src="https://nextcloud.com/s/1234/preview">
```
It'll load and display without issues.

**Comment by fschrempf** [comment]
Is there a way for whitelisting a specific origin from which I want to access a public calendar link like `https://nextcloud.example.com/apps/calendar/p/xyz1234` without needing user authentication?

If not I think there should be one. If I have a website/webapp that I'm controlling myself I want to have the possibility to let it access NC data without running into CORS issues.

**Comment by photopea** [comment]
Guys, could you simply always enable CORS by default in all Nextcloud instances?

Authors of any webapp can always create a server-side "mirroring" of the HTTP requests, and their server does not care about the CORS. So not having CORS headers does not add any security, it only makes everything more complicated.

**Comment by salonikumawat28** [comment]
I looked into similar issues like https://github.com/nextcloud/server/issues/3131 and all the solution attempts and temporary hacks I did were based on the suggested based in the issue, NextCloud [login v2 flow](https://docs.nextcloud.com/server/stable/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2) doc and NextCloud [OAuth flow](https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/oauth2.htm) doc but I am not able to find a solution which doesn't involve changes in NextCloud server. Is there any way I can solve CORS issues be config changes and  may be by writing a NextCloud app?

**Comment by pbek** [comment]
There still is https://github.com/digital-blueprint/webapppassword.

**Comment by salonikumawat28** [comment]
Hello everyone, I have created a NextCloud app which adds a file menu item in `Files` app. When you click this menu item, it opens my website. From this website, I want to login into NextCloud so that I can get credentials which I can use to download and upload files. 

First step I did was using the hardcoded `username:password` and passed it in Authorization header to download and upload URLs which failed on CORS. I was able to resolve the CORS issues once I installed WebAppPassword and whitelisted my website.

As a second step, I replaced the hardcoded `username:password` with the [login v2 flow](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2) and it failed on CORS. This was surprising as my website origin was already whitelisted in WebAppPassword and was working for download and upload URLs. I also tried [OAuth2 flow](https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/oauth2.html) but it also failed on CORS and preflight.

Is there a way to avoid CORS and preflight issues in login v2 flow or OAuth flow?
You can find more details at this issue I created: https://github.com/nextcloud/server/issues/34898

**Comment by bonnebulle** [comment]
> There still is https://github.com/digital-blueprint/webapppassword.

Yes, the android app say to add "https://app.super-productivity.com", it is the same with linux/desktop app ?
Thanks (here, is not working)

**Comment by pbek** [comment]
That's a great workaround!

**Comment by apfohl** [comment]
> That's a great workaround!

The problem is, that I as a client developer can not rely on having this in place for random Nextcloud instances.

**Comment by pbek** [comment]
...nor random Nextcloud Apps for that matter.

**Comment by apfohl** [comment]
> ...nor random Nextcloud Apps for that matter.

The Apps have probably different methods for pushing files. But idk.

**Comment by pbek** [comment]
I meant "the permissions to install an app that sets headers", like https://github.com/digital-blueprint/webapppassword. 😉

**Comment by jthoward64** [comment]
I don't understand why simply allowing access for [login v2](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2) is such an issue, NextCloud already rejects unauthorized requests, so setting CORS for the WebDAV gateway shouldn't be an issue either, but if that is the blocker oh well, SSR web apps can get around that easily enough. The main issue for web apps, as far as I can tell, is the `/login/v2` route. If that got support (even if it behind a config file flag, though not preferable) then it would open up a world of opportunities. And I doubt anyone is going to be using my Nextcloud for a CDN for my login page.

**Comment by mmuman** [comment]
I have a similar problem with using the download link for a markdown file from a js app on a separate domain: the default CORS policy blocks it. I don't see any reason those would be… Should I open a separate issue?

**Comment by pkuegler** [comment]
Solution:
1. Install "https://github.com/digital-blueprint/webapppassword" on the nextcloud instance you want to access from a web-client
2. In Administration Settings/WebAppPassword add the source domain as allowed origin for webdav/caldav
3. Use some web-based code to access the resource. Make sure that the preflight is without authantication

Example:
If you want to retrieve a file listing on your webserver (webserver.somedomain.com) from your nextcloud (nextcloud.yourdomain.com) do as follows:
1. Add "https://webserver.somedomain.com" as allowed origin in nextcloud/WebAppPassword
2. Use e.g. something lioke this code on your website:

fetchDataFromWebDav(baseUrl, folderPath)
        .then(parseXmlResponse)
        .then(data => renderFileList(data, baseUrl))
        .catch(error => {
            console.error('Error:', error);
            widgetContainer.innerHTML = '<p>Error loading directory contents.</p>';
        });


function fetchDataFromWebDav(baseUrl, folderPath) {
    const fullUrl = `${baseUrl}${encodeURIComponent(folderPath)}`;
    const username = 'testuser';
    const password = 'testuserpwd'; // just for show - use SSO!
    const headers = new Headers({
        'Authorization': 'Basic ' + btoa(username + ":" + password),
        'Content-Type': 'application/xml; charset=utf-8'
    });

    return fetch(fullUrl, { method: 'PROPFIND', headers });
}

function parseXmlResponse(response) {
    if (!response.ok) {
        throw new Error('Network response was not ok');
    }
    return response.text()
        .then(str => (new window.DOMParser()).parseFromString(str, "text/xml"));
}


function renderFileList(data, baseUrl) {
    const files = data.querySelectorAll('response');
    let fileList = '<ul>';

    files.forEach(file => {
        const href = file.querySelector('href').textContent;
        const name = decodeURIComponent(href.split('/').pop());
        if (!name) return; // Skip empty names

        const isFolder = file.querySelector('resourcetype').querySelector('collection') !== null;
        const icon = isFolder ? '📁' : '📄';
        const itemPath = href.replace('/nextcloud/remote.php/dav/files/testuser', ''); // Adjust path as needed

        fileList += `<li>${icon} <a href="${href}" class="${isFolder ? 'folder-link' : ''}" data-path="${href}">${name}</a></li>`;
    });

    fileList += '</ul>';
    // Do something with fileList
}

**Comment by pkuegler** [comment]
### Does it work for public shares?
Unfortunately not. The above solution only works for "private" shares that are being accessed via "^/remote.php/dav/files...". If you also want to access public shares "^/public.php/webdav/...", then the WebAppPassword does nto help you, as it does not take care of the cors headers there. You can see that, because all preflight requests (done by the browser automatically before every fetch) will fail because of "missing basic authentication header". And just to be clear: This is a nextcloud bug, as based on specs, there **must not be** authentication headers in the preflight options request.

### Can it work for public shares?
Yes, it can, but you need to modify the .htaccess for this (in the nextcloud root folder). There should be already a couple of rewrites there ;-) If you are using a reverse proxy, you can add the rewrites there as well, even though that might me more ressource-intensive, as the proxy needs to cache the outgoing response to be able to rewrite headers (you can use nginX/LUA for that).

**Just add following code (adjust to your paths etc.) to the .htaccess:**
```
<IfModule mod_headers.c>
    # Apply to all requests
    Header always set Access-Control-Allow-Origin "https://webserver.somedomain.com"
    Header always set Access-Control-Allow-Credentials "true"
    Header always set Access-Control-Allow-Methods "PROPFIND"
    Header always set Access-Control-Allow-Headers "authorization,content-type"

    # Specific handling for OPTIONS requests
    <If "%{REQUEST_METHOD} == 'OPTIONS'">
        SetEnvIf Origin "https://webserver.somedomain.com" PREFLIGHT_REQUEST_FROM_WEBSITE
        Header merge Access-Control-Allow-Origin "https://webserver.somedomain.com" env=PREFLIGHT_REQUEST_FROM_WEBSITE
        Header merge Access-Control-Allow-Methods "PROPFIND" env=PREFLIGHT_REQUEST_FROM_WEBSITE
        Header merge Access-Control-Allow-Headers "authorization,content-type" env=PREFLIGHT_REQUEST_FROM_WEBSITE
        Header merge Access-Control-Allow-Credentials "true" env=PREFLIGHT_REQUEST_FROM_WEBSITE
        # Respond with 204 No Content for OPTIONS requests
        # Note: This directive might not be respected depending on the server configuration and Apache version
        RewriteEngine On
        RewriteCond %{REQUEST_METHOD} OPTIONS
        RewriteRule .* - [R=204,L]
    </If>
</IfModule>
```

This enforces/overwrite the headers to tell the broser to move forward with the actual fetch request. Kepp in mind, that this is not adding any security risk, as the actual request (PROPFIND) will be checked correctly for the authorization headers. THe actual problem is, that the Browser nevery tries this request, as the preflight failes due to this Nextcloud bug.

Have fun!

**Comment by joshyy1312** [comment]
I tried the htaccess rewrite to use a public share but the PROPFIND request still fails with a CORS error (OPTION passes with 204)

Am I correct to use "myNextCloud/public.php/webdav" with the share token as user and an empty password?

**Comment by joshyy1312** [comment]
It still didn't work.
With your htaccess code the **OPTIONS** request works and then the **PROPFIND** fails with a **CORS** error. 

In my case the file is a public share without password, it just needs the token as login and an empty password. Everything is correct because I can see the **XML** file listing in the browser **XHR** console but it just isn't passed to my script.

From what I see in the browser inspector, with the **PROPFIND** request, the "always set" header lines are not set.

**Fortunately, here is the solution I found and it works:**

_It **might** be related to my hosting provider and an apache/fastcgi bug as stated [here](https://stackoverflow.com/questions/53259981/htaccess-headers-being-ignored-by-apache)._

Using one of the solution in the above article, namely adding the same headers through php directly, worked for me.

There is a file called **.user.ini** in the hosting root and it allows to set php properties or settings.

If you add the following line it will prepend a php file to every php file request.
```
auto_prepend_file="myheaders.php"
```

The **myheaders.php** file looks like this:

```
<?php
header('Access-Control-Allow-Origin: https://your.other.domain');
header('Access-Control-Allow-Credentials: true');
header('Access-Control-Allow-Methods: PROPFIND, GET');
header('Access-Control-Allow-Headers: authorization,content-type');
?>
```

It now seems to work without problem, I can retrieve the file listing and also GET a single file directly.

Many thanks **pkuegler** for pointing me in the right direction and taking the time to explain it in detail, I've been looking for weeks about a way to achieve this!

**Comment by david-haerer** [comment]
> Also interested in this.
> 
> If others are stuck but use NextCloud behind nginx as a reverse proxy I can recommend https://www.williamjbowman.com/blog/2021/05/13/enabling-cors-for-nginx-webdav-and-caldav-reverse-proxy/ which worked for me for https://github.com/bfren/docker-nginx-webdav
> 
> [@photopea](https://github.com/photopea) do you mean something like https://github.com/sebastienvercammen/node-cors-proxy-server and if not do you have a recommendation until this is implemented?

For anyone running Nextcloud behind Caddy, this is how I got it working.
The trick is respond to the CORS preflight request from `myapp.example.com` with Caddy,
since Nextcloud would respond with `405 Method Not Allowed`.
The config might need some adaption if you want to do other requests than `PUT` (for file upload).

```Caddyfile
nextcloud.example.com {
	# Matcher for CORS preflight request from my app.
	@myapp {
		method OPTIONS
		header Origin https://myapp.example.com
	}

	# Set CORS response headers for the preflight request.
	header @myapp Access-Control-Allow-Origin "https://myapp.example.com"
	header @myapp Access-Control-Allow-Methods "PUT"
	header @myapp Access-Control-Allow-Headers "content-type"

	# Respond successfully the preflight request.
	respond @myapp 204
	
	reverse_proxy 127.0.0.1:11000 {
		# Optionally set the CORS headers in the Nextcloud response.
		header_down Access-Control-Allow-Origin "https://myapp.example.com"
		header_down Access-Control-Allow-Methods "PUT"
		header_down Access-Control-Allow-Headers "content-type"
	}
}
```

**Comment by coderofsalvation** [comment]
Is it possible that this has been resolved?
I've installed Nextcloud Hub 9  (30.0.10) some months ago (hetzner's storageshare-tier), and I was able to perform all curl request (including PROPFIND) from [the docs](https://docs.nextcloud.com/server/30/user_manual/en/files/access_webdav.html) without any issue.

My feeling is that this issue is conflating reverse proxy-configuration with nextcloud configuration.

If the documentation even demonstrates WebDav curl calls, we should first assume the issue is not in nextcloud.
Rather, it's probably (mis)configuration of whatever is **in front of nextcloud** (reverse proxy e.g.).
Keeping this issue open makes it looks like nextcloud has webdav issues, without actually having them :)

**Comment by perry-mitchell** [comment]
CORS is the responsibility of the application, so Nextcloud should be setting those headers. It should indeed not be the responsibility of whatever downstream reverse proxy configuration may or may not be intercepting requests. I suspect most configurations would simply return the headers output by Nextcloud, so if Nextcloud didn't specify them, then it was Nextcloud itself that is to blame, imo.

It certainly doesn't _help_ that there are tonnes of online guides that involve setting up a reverse proxy, and I agree that will absolutely be a source of confusion. My point here is that if Nextcloud is finally outputting CORS headers directly from the server itself, then there's a high likelihood that it should be fixed.

@coderofsalvation Can you confirm that you're seeing CORS headers in responses from Nextcloud? Just saying it works won't confirm that CORS is present - and it indeed needs to be present for this issue to be closed, imo.

I'm running a similar version so I can also check a bit later as well.

EDIT: Additionally, for the scope of this issue to have been addressed, said CORS headers need to allow for either:
 * Allow-listed domains _or some other fine-grain control_
 * Global-allow list (`*`)
 
And I'm not sure if the Nextcloud UI has these options. CORS is a security feature and should allow for some control over who can request resources from a browser-based environment.

**Comment by coderofsalvation** [comment]
> @coderofsalvation Can you confirm that you're seeing CORS headers in responses from Nextcloud?

I did a grep on `Access-Control` and did not get a match. I guess we can conclude it's not there (and `curl` does not care).

hold on, I will try the same call via a `fetch()` request in the browser for v30.0.10

> UPDATE: aaaaand it fails in the browser due to content security policy violation.

Ok, agreed.
So the next step forward is a PR with @joshyy1312 's patch?
I have a hunch we need to set those headers conditionally (for `/remote.php/dav/*` URLs only) to pass a security audit.

> For the time being, I guess I have to run `rclone serve` to re-serve nextcloud webdav  **with the headers**.

**Comment by perry-mitchell** [comment]
Just tried on `Nextcloud Hub 9 (30.0.6)`, and I don't see any CORS headers unfortunately:

```
HTTP/1.1 207 Multi-Status
Date: Wed, 11 Jun 2025 10:27:54 GMT
Content-Type: application/xml; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Referrer-Policy: no-referrer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
X-Robots-Tag: noindex, nofollow
X-XSS-Protection: 1; mode=block
X-Powered-By: PHP/8.2.27
Set-Cookie: ocfhupkilpx4=<snip>; path=/; HttpOnly; SameSite=Lax
Set-Cookie: oc_sessionPassphrase=<snip>; path=/; HttpOnly; SameSite=Lax
Set-Cookie: ocfhupkilpx4=<snip>; path=/; HttpOnly; SameSite=Lax
Content-Security-Policy: default-src 'none';
Set-Cookie: nc_sameSiteCookielax=true; path=/; httponly;expires=Fri, 31-Dec-2100 23:59:59 GMT; SameSite=lax
Set-Cookie: nc_sameSiteCookiestrict=true; path=/; httponly;expires=Fri, 31-Dec-2100 23:59:59 GMT; SameSite=strict
Set-Cookie: ocfhupkilpx4=<snip>; path=/; HttpOnly; SameSite=Lax
Set-Cookie: ocfhupkilpx4=<snip>; path=/; HttpOnly; SameSite=Lax
Set-Cookie: ocfhupkilpx4=<snip>; path=/; HttpOnly; SameSite=Lax
Vary: Brief,Prefer
DAV: 1, 3, extended-mkcol, access-control, calendarserver-principal-property-search, nextcloud-checksum-update, nc-calendar-search, nc-enable-birthday-calendar
X-Request-Id: <snip>
X-Debug-Token: <snip>
```

I'm aware that I'm not on the latest version, but I haven't seen anything that would indicate them having been added yet.

**Comment by sunjam** [comment]
@pbek curious what you think about status of this longtime issue.  Your app has being discussed here and it has seen significant development!

**Comment by pbek** [comment]
I don't know whether there is a reason NOT to put CORS handling into Nextcloud server directly. ️🤷🏻 
WebAppPassword was never meant to handle CORS for everything. 😊 But since I put a nice Nix powered CI with VMs in place to test if the sharing API breaks the whole instance last month, I guess it will stay there for a while...

**Comment by joshyy1312** [comment]
Hi,

Is this still planned to be worked on someday?

Because the workaround I proposed, while it works and I still use it, breaks some of the standard nextcloud functionalities.

In my case it's usable because I just use that specific nextcloud instance to host the files I need in the other website. But I have to 'disable' it every time I want to add a new file so I can setup the public share in the nextcloud interface.

--

To give some details, I use nextcloud to host/sync my obsidian vault across multiple devices.
And I also use nextcloud to share specific notes and embed their content in a website. That way I dont have to duplicate the content both in obsidian and the site's backend. The notes kinda serve as the backend.

In any case, many thanks for nextcoud :)

**Comment by mabuch** [comment]
Hi @joshyy1312 
I cannot answer your question, but I want to tell you about another software I am now using (in parallel to Nextcloud) for some synchronising purposes. I am Using PicoDAV, which is a very lightweight WebDAV implementation which perfectly suits some of my needs, where I simply need to synchronise some files. Maybe this would work for Obsidian as well. If you need a Nextcloud compatible Webdav implementation, have a look at KaraDAV, the more feature-rich version of PicoDAV (I have not tried KaraDAV myself, as PicoDAV does all I need).
I still use Nextcloud for the management of my files, but for another purpuse where I simply needed a WebDAV backend to sync files for another software, I have switched to PicoDAV. This might not be suitable for you, or I can understand if you don't want to introduce another tool (as it is another software which needs to be kept up to date), but for my purpose it works very well. Although I would also have prefered if it would have been possible to only use Nextcloud for everything.

**Comment by Utopiah** [comment]
Sad to come to this point but in terms of lightweight easy to setup alternatives I can also recommend CopyParty. It is "just" a fileserver with only a markdown editor (and a lot more actually) without a complex plugin system or and ecosystem... but it works and with CORS optionally.

**Comment by susnux** [comment]
This feature is on our schedule but we cannot yet give an estimation.

---

Instead of a second server you can still use  https://apps.nextcloud.com/apps/webapppassword for this.

**Comment by coderofsalvation** [comment]
@Utopiah thanks for the suggestion! :)
@susnux thanks for the suggestion too. Could you get it to work? for example with https://diffuse.sh?
I've installed [webapppassword](https://apps.nextcloud.com/apps/webapppassword) on my hetzner (yourstorage, managed nextcloud) and added 'https://diffuse.sh' to the webdav origins-textfield, and added a webdav app password (in security tab).
When adding the webdav credentials to (click 'add music') it errors with nextcloud replying 'Cannot connect to this source'.
I think a musicplayer is a perfect usecase for how nextcloud's webdav can share data with other apps.

**Comment by pbek** [comment]
@coderofsalvation did you configure the origins in webapppassword?

**Comment by coderofsalvation** [comment]
yes

**Comment by coderofsalvation** [comment]
@susnux @pbek were you able to get webapppassword going with webdav-webapps like [diffuse.sh](https://diffuse.sh)?

**Comment by pbek** [comment]
I never tried that.
